shared/src/Datadog.Trace.ClrProfiler.Native/crashhandler.cpp

```diff
@@ -0,0 +1,409 @@
+
+        // If we successfully unregistered the .NET handler, put it back in place
+        if (SUCCEEDED(unregisterDacHr))
+        {
+            auto registrationHr2 = WerRegisterRuntimeExceptionModule(dacFilePath.c_str(), (PVOID)hModule);
+
+            if (FAILED(registrationHr2))
+            {
+                Log::Warn("Crashtracking - Failed to re-register the DAC handler: ", registrationHr2);
+            }
+        }
+
+        if (FAILED(registrationHr))
+        {
+            Log::Warn("Crashtracking - Could not register the crash handler: ", registrationHr);
+            return false;
+        }
+
+        _crashHandler = dllPath;
+        return true;
+    }
+
+    extern "C"
+    {
+        HRESULT __declspec(dllexport) OutOfProcessExceptionEventCallback(
+            PVOID pContext,
+            const PWER_RUNTIME_EXCEPTION_INFORMATION pExceptionInformation,
+            BOOL* pbOwnershipClaimed,
+            PWSTR pwszEventName,
+            PDWORD pchSize,
+            PDWORD pdwSignatureCount
+        )
+        {
+            if (pbOwnershipClaimed != nullptr)
+            {
+                // We don't claim the ownership of the crash.
+                // This way, the original crash handler registered by .NET will be invoked,
+                // and we don't affect the normal behavior.
+                *pbOwnershipClaimed = FALSE;
+            }
+
+            // Get the pid and tid from the exception
+            auto pid = GetProcessId(pExceptionInformation->hProcess);
+            auto tid = GetThreadId(pExceptionInformation->hThread);
+
+            // Read the environment variables saved in the crashing process
+            WerContext context{};
+
+            BOOL hasContext = ReadProcessMemory(pExceptionInformation->hProcess, pContext, &context, sizeof(context), nullptr);
+            BOOL hasEnviron = FALSE;
+
+            std::vector<WCHAR> envBlock;
+
+            if (hasContext && context.EnvironLength > 0 && context.Environ != nullptr)
+            {
+                envBlock.resize(context.EnvironLength * sizeof(WCHAR));
+                hasEnviron = ReadProcessMemory(pExceptionInformation->hProcess, context.Environ, envBlock.data(), context.EnvironLength * sizeof(WCHAR), nullptr);
+            }
+
+            // Merge them with the current environment variables
+            auto currentEnv = GetEnvironmentStrings();
+
+            if (hasEnviron)
+            {
+                envBlock = MergeEnvironmentBlocks((LPCWSTR)envBlock.data(), currentEnv);
+            }
+            else
+            {
+                envBlock.assign(currentEnv, currentEnv + envBlock.size());
+            }
+
+            FreeEnvironmentStrings(currentEnv);
+
+            // Create the command-line for dd-dotnet
+            std::filesystem::path p(GetCurrentDllPath());
+            auto directory = p.parent_path();
+            auto ddDotnetPath = directory / "dd-dotnet.exe";
```

Comment:
Do we need to check if this actually exists? If not I assume the CreateProcess just fails and we bail, but there's no errors or anything logged about that?

Reply:
Good point